offload/new-api/include/offload_api.h

```diff
@@ -0,0 +1,415 @@
+///         + `NumEntries > 0 && phDevices == NULL`
+///     - ::OFFLOAD_RESULT_ERROR_INVALID_VALUE
+///     - ::OFFLOAD_RESULT_ERROR_INVALID_NULL_HANDLE
+///         + `NULL == hPlatform`
+OFFLOAD_APIEXPORT offload_result_t OFFLOAD_APICALL offloadDeviceGet(
+    // [in] handle of the platform instance
+    offload_platform_handle_t hPlatform,
+    // [in] the type of the devices.
+    offload_device_type_t DeviceType,
+    // [in] the number of devices to be added to phDevices. If phDevices is not
+    // NULL, then NumEntries should be greater than zero. Otherwise
+    // OFFLOAD_RESULT_ERROR_INVALID_SIZE will be returned.
+    uint32_t NumEntries,
+    // [out][optional] Array of device handles. If NumEntries is less than the
+    // number of devices available, then platform shall only retrieve that
+    // number of devices.
+    offload_device_handle_t *phDevices,
+    // [out][optional] pointer to the number of devices. pNumDevices will be
+    // updated with the total number of devices available.
+    uint32_t *pNumDevices);
+
+///////////////////////////////////////////////////////////////////////////////
+/// @brief Retrieves various information about device
+///
+/// @details
+///    - The application may call this function from simultaneous threads.
+///    - The implementation of this function should be lock-free.
+///
+/// @returns
+///     - ::OFFLOAD_RESULT_SUCCESS
+///     - ::OFFLOAD_RESULT_ERROR_UNINITIALIZED
+///     - ::OFFLOAD_RESULT_ERROR_DEVICE_LOST
+///     - ::OFFLOAD_RESULT_ERROR_UNSUPPORTED_ENUMERATION
+///         + If `propName` is not supported by the adapter.
+///     - ::OFFLOAD_RESULT_ERROR_INVALID_SIZE
+///         + `propSize == 0 && pPropValue != NULL`
+///         + If `propSize` is less than the real number of bytes needed to
+///         return the info.
+///     - ::OFFLOAD_RESULT_ERROR_INVALID_NULL_POINTER
+///         + `propSize != 0 && pPropValue == NULL`
+///         + `pPropValue == NULL && pPropSizeRet == NULL`
+///     - ::OFFLOAD_RESULT_ERROR_INVALID_DEVICE
+///     - ::OFFLOAD_RESULT_ERROR_OUT_OF_RESOURCES
+///     - ::OFFLOAD_RESULT_ERROR_OUT_OF_HOST_MEMORY
+///     - ::OFFLOAD_RESULT_ERROR_INVALID_NULL_HANDLE
+///         + `NULL == hDevice`
+OFFLOAD_APIEXPORT offload_result_t OFFLOAD_APICALL offloadDeviceGetInfo(
+    // [in] handle of the device instance
+    offload_device_handle_t hDevice,
+    // [in] type of the info to retrieve
+    offload_device_info_t propName,
+    // [in] the number of bytes pointed to by pPropValue.
+    size_t propSize,
+    // [out][optional] array of bytes holding the info. If propSize is not equal
+    // to or greater than the real number of bytes needed to return the info
+    // then the OFFLOAD_RESULT_ERROR_INVALID_SIZE error is returned and
+    // pPropValue is not used.
+    void *pPropValue,
+    // [out][optional] pointer to the actual size in bytes of the queried
+    // propName.
+    size_t *pPropSizeRet);
```

Comment:
I think the very first thing we should have implemented is a proper reference counted `init/shutdown` function pair. We probably not need to initialize plugins until the user tries to get a device though.

Reply:
I've added `offloadInit` (which initializes the plugins) and `offloadShutDown` (which does nothing at the moment)